### DISCUSSION TITLE
High Availability using Static files or ISR files

### DISCUSSION THREAD

**Discussion by leroy0211** [vercel/next.js#22526] [Help]
I've been thinking about High Availability and NextJs. 

When you have a setup in which you have multiple servers running your NextJS application, how can you be sure all static (pregenerated) pages with or without ISR (Incremental Static Regeneration) are using the same data. 

Lets assume a large e-commerce site with 500.000 products, with quickly changing prices, stock details, etc. And increasing product list on a daily basis. 
And let's say you deploy the NextJS application to many instances (because you want to be High Available, and can process all the load). Let's say you use kubernetes with an autoscaler. So replica's can be as much as you need. 

When each NextJS application stores pre-rendered pages (with or without ISR) on it's own filesystem, how could you be sure that every instance is serving the latest information? 

And another question I've been thinking of; is it even a good practice to pre-render (using ISR) near 500.000 pages? 
Of-course, not during build time, but during runtime, using `fallback: true`. 

The only thing I personally can think of is to use `fallback: true` and a `revalidate: 1` (re-render every 1 sec.). And use an empty `getStaticPaths` function. So during build time, none of the pages are pre-rendered.

What do you people think of this situation? Am I missing something, is there some static page caching mechanism I've overlooked?

**Comment by smontlouis** [comment]
From what I understand with revalidate, when you use pre-rendered pages, at some point in time there will always be that user who will see the page with outdated information. His arrival on the page will trigger the revalidate to refresh the information for the next user.

A solution might be to use revalidate at 1  like you said and trigger a webhook to that specific page in order to update it before any user can reach it.

Regarding your second question, indeed you will save some query costs. It's better to have one query per second than one query per user visit.

[Seems like this convo might help you ](https://github.com/vercel/next.js/discussions/11552#discussioncomment-52084)

**Reply by leroy0211** [reply]
> A solution might be to use revalidate at 1 like you said and trigger a webhook to that specific page in order to update it before any user can reach it.

This is quite cumbersome for my use-case of 500.000+ (Product) pages. I think in total we have around 1 million pages (non user specific)

> Regarding your second question, indeed you will save some query costs. It's better to have one query per second than one query per user visit.

It's not the query cost I'm worried about, it's more like storage and correctness in a High Availability environment with multiple replica's that I'm worried about. 

Let's say I have 6 replica's of the application, and the load balancer does round-robin. 

If each pre-rendered page is like around 50kb, to pre-render around 500.000 pages, this would take up to to 25GB's of storage per replica. Which need to be re-rendered on every new deployment. Because on each deployment of the application those pre-rendered pages are gone.

If each replica is storing pre-rendered pages, but not sharing the pages. It's possible that each replica have different information. And yes, with a `revalidate: 1` we probably could limit this problem,  but don't get rid of it. 

It would be great if those pre-rendered pages could be shared somehow, or cached in Redis or something, so they can be shared with replica's, but also re-used cross deployments of the application.

**Reply by reckter** [reply]
If you do not pre-render the pages at build-time (with dynamic urls and an empty `getStaticPaths` function for example. Your cache will grow with time (and usage) and not be filled at build-time. This does not completely alleviate the problem, but it might help. (BUT: you will have to deal with the first rendering, either by having a `fallback: blocking` and slower load time, or with a fallback page).

If you can switch to vercel.com (or another serverless hosting solution for next.js applications, not sure if there are any out there, but maybe), you will no longer have that problem at all. The cache will be global, and shared world wide by the vercel.com CDN. (Not affiliated with vercel at all, but I really like the vercel.com nextjs integration). From personal experience build time will also be way faster. We went from a ~20 minute rollover on kubernetes (3 pods, rolling deployment), to a mere ~3 minute build time. (+ added benefit DX increased X-fold, because of preview deployments, with the right setups those do magic for QA!)

I honesty wouldn't worry about the correctness down to the second. (The way the next.js caching works you will ALWAYS have stale data on some users). 

If you need those kinds of correctness, SSG is not the right tool, and you should switch to SSR instead.
I challenged the second-correctness in our application as well, and we went up to a 10-minute cache time on most pages. YMMW obviously, but for most applications these numbers are probably good enough. (user-specific pages excluded of course!)

**Reply by stigkj** [reply]
It is impossible to have 100% correctness when you have multiple instances, no matter if it is SSG or SSR. But if you use ISR (Incremental Static Regeneration) with a CDN in front, you will get very close. And then redeployment and storage won't be an issue. The CDN does not have to be Vercel, but that is of course the easiest integration.

**Reply by leroy0211** [reply]
@reckter thanks for the tip! I've already started looking into an empty `getStaticPaths` and `getStaticProps` solution. So builds can still be quick. I haven't spent much time on the `fallback` property. But to my understanding it's either wait for the page to load (which can cause timeouts). Or have some  `Loading...` page, so the visitor sees that things are loading.


Hosting on vercel, sadly, is a no-go for the Org. I'm working at, at this moment. It's an ethical thing.

I've been looking at something like [next-redis-cache](https://github.com/knovator/next-redis-cache) the last few days. Seems like that packages deliveres what we require. A shared page cache (pre-render at runtime, not build-time), so every nextjs application instance can serve the same version of the information on the page. I still need to test it though. 

The correctness of our pages are actually one of the top priorities. It reduces phone-calls, e-mails, etc. to our customer support and others, and therefor reduces costs. There are quite some use-cases where customers call, because there is a price difference of merely $0,10 between product page vs. checkout. 

I'm not one of the main developers on this project in our Org. but I'm also thinking that some pages are required to do SSR instead of SSG. Right now none of the pages do SSR or SSG, they all load their data from within the browser, calling api's. Which works, but doesn't work well with SEO.

**Reply by stigkj** [reply]
@leroy0211 Regarding the empty `getStaticPaths/getStaticProps` solution, your understanding is correct. Either let it be `blocking` so the page is not rendered before `getStaticProps` has returned data, or have a loading version of the page. The loading version can be an almost complete page, i.e. it is just the parts that need dynamic data that needs to have loading indicators.

If you absolutely need the correct prices, you cannot cache these. To still have fast pages, you can do SSG for most of the page and just load the prices from the client. This also lets you have good SEO. As you have a lot of pages, you need to choose a list of pages that should be rendered on build and let the rest render on demand through the `fallback` support.

tl;dr:

```
export async function getStaticPaths() {
    return {
        paths: <list of the most important product pages to prerender on build>,
        fallback: true, // i.e. all paths not in this list will render a fallback page on first request
    };
}

export async function getStaticProps({ params }) {
    const data = <fetch data for params>

    return {
        props: {
          data,
        },
        // Updates the cached pages after x seconds. Use this if the product pages
        // has something other than prices that needs to be updated from time to time
        revalidate: 60,
    }
}

const ProductPage = ({ data }) => (
    const router = useRouter()

    useEffect(() => {
        ...code to fetch prices in the browser...
    }, [<some deps>])
     
    if (router.isFallback) {
        ...code to render product page without data...
        // shortly after, getStaticProps is finished and the page is rendered with data
    } else {
        ...code to render full product page without prices...
    } 

);

export default ProductPage;
```

**Reply by stigkj** [reply]
More on SEO: remember that you have a crawl budget, i.e. Google won't have the time to index all your pages. So if you let your most important product pages be in the list returned from `getStaticPaths` and make them indexable, you could turn on noindex on the rest of the product pages.

**Reply by leroy0211** [reply]
@stigkj I know google has crawl budget, but eventually all marketable pages need to be indexed. Our Marketing department knows best about that. Having product pages with `noindex` means that google never knows about them. That's like shooting in your own foot for a company with over 120 million in revenue. 

Pre-generating pages is also hard, because it will take hours during build time. I think the best approach here is to use ISR with a very short time for dynamic pages.  But what about normal pages like the  `pages/product/index.js`? There is no ISR possible, only Static or SSG, right? 


I'm also thinking about other ways of having my cache stored elsewhere, like varnish page cache or redis, combining it with ESI rendering for content that almost never changes, like menu's, footers, headers, etc.

**Reply by stigkj** [reply]
@leroy0211 Not sure if you have seen it, but Next.js has added support for crawlers: https://nextjs.org/docs/basic-features/data-fetching#fallback-true.

tl;dr: for a user, the missing data is fetched client-side, while for crawlers it will do it server-side, i.e. it will behave as if `fallback: blocking`.

Regarding normal pages you could use static rendering on build as I guess you won't have that many of them? You could do ISR for them too, but then I think you would need to have some `[]` in the filename/directories.

**Reply by leroy0211** [reply]
It's still not exactly usable for our use-case. BTW, we stopped working on nextjs, but personally I would still like to explore this a bit more. 

It's still using the `getStaticPaths`, which can only work with over 5 million pages to render if you add `[]` as paths list.  The problem then only rises with urls without parameters, like `/products` instead of `/products/{id}`. 

What we thought NextJS could do for us, is to render any type of page during runtime (not build time). And have some central cache in place for static rendered pages, so that a high available self-hosted setup could also benefit from the ISR or any other static rendered pages. Also during a redeployment of the application, so that ISR pages are still static.
